Source/WebKit/UIProcess/ProvisionalPageProxy.cpp

```diff
@@ -198,36 +196,44 @@ void ProvisionalPageProxy::initializeWebPage(RefPtr<API::WebsitePolicies>&& webs
     m_drawingArea = m_page->pageClient().createDrawingAreaProxy(m_process.copyRef());
 
     bool sendPageCreationParameters { true };
+    bool registerWithInspectorController { true };
     auto parameters = m_page->creationParameters(m_process, *m_drawingArea, WTFMove(websitePolicies));
 
     if (page().preferences().processSwapOnCrossSiteWindowOpenEnabled() || page().preferences().siteIsolationEnabled()) {
+        // FIXME: <rdar://121241026> This is not elegant and not robust. It needs to be.
         RegistrableDomain navigationDomain(m_request.url());
         RefPtr openerFrame = m_page->openerFrame();
-        if (RefPtr openerPage = openerFrame ? openerFrame->page() : nullptr) {
+        RefPtr openerPage = openerFrame ? openerFrame->page() : nullptr;
+        if (openerFrame) {
             parameters.openerFrameIdentifier = openerFrame->frameID();
             parameters.mainFrameIdentifier = m_page->mainFrame()->frameID();
-
-            if (auto existingRemotePageProxy = m_page->takeRemotePageProxyInOpenerProcessIfDomainEquals(navigationDomain)) {
-                ASSERT(existingRemotePageProxy->process().processID() == m_process->processID());
-                m_webPageID = existingRemotePageProxy->pageID();
-                m_mainFrame = existingRemotePageProxy->page()->mainFrame();
-                m_messageReceiverRegistration.stopReceivingMessages();
-                m_messageReceiverRegistration.transferMessageReceivingFrom(existingRemotePageProxy->messageReceiverRegistration(), *this);
-                LocalFrameCreationParameters localFrameCreationParameters {
-                    std::nullopt
-                };
-                m_process->send(Messages::WebPage::TransitionFrameToLocal(localFrameCreationParameters, m_page->mainFrame()->frameID()), m_webPageID);
-                sendPageCreationParameters = false;
-            } else if (RefPtr existingRemotePageProxy = openerPage->remotePageProxyForRegistrableDomain(navigationDomain)) {
-                ASSERT(existingRemotePageProxy->process().processID() == m_process->processID());
-                openerPage->addOpenedRemotePageProxy(m_page->identifier(), existingRemotePageProxy.releaseNonNull());
-                m_needsCookieAccessAddedInNetworkProcess = true;
-            } else {
-                auto remotePageProxy = RemotePageProxy::create(*openerPage, m_process, navigationDomain);
-                remotePageProxy->injectPageIntoNewProcess();
-                openerPage->addOpenedRemotePageProxy(m_page->identifier(), WTFMove(remotePageProxy));
-            }
         }
+        auto existingRemotePageProxy = m_page->takeRemotePageProxyInOpenerProcessIfDomainEquals(navigationDomain);
+        if (!existingRemotePageProxy) {
+            if ((existingRemotePageProxy = m_page->takeOpenedRemotePageProxyIfDomainEquals(navigationDomain)))
```

Comment:
There are extra parentheses here.

Reply:
Fun fact: these are not actually unnecessary parentheses.  They are needed to prevent the compiler warning me that I am checking the nullness of the result of the assignment operator, not checking equality of two objects.